plenum/common/stacks.py

```diff
@@ -34,14 +37,70 @@ def __init__(self, stackParams: dict, msgHandler: Callable, seed=None,
             seed=seed,
             onlyListener=True,
             config=config,
-            msgRejectHandler=msgRejectHandler)
+            msgRejectHandler=msgRejectHandler,
+            create_listener_monitor=True)
         MessageProcessor.__init__(self, allowDictOnly=False)
-        self.connectedClients = set()
+
+        self.max_connected_clients_num = config.MAX_CONNECTED_CLIENTS_NUM
+        self.postrestart_wait_time = config.STACK_POSTRESTART_WAIT_TIME
+        self.min_stack_restart_timeout = config.MIN_STACK_RESTART_TIMEOUT
+        self.max_stack_restart_time_deviation = config.MAX_STACK_RESTART_TIME_DEVIATION
+        self.init_stack_restart_params()
+
+    def init_stack_restart_params(self):
+        self.connected_clients_num = 0
+        self.stack_restart_is_needed = False
+        self.last_start_time = time.time()
+        self.next_restart_min_time = self.last_start_time + \
+            self.min_stack_restart_timeout + \
+            randint(0, self.max_stack_restart_time_deviation)
+
+    def check_listener_events(self):
+        events = self.get_monitor_events(self.listener_monitor)
+        for event in events:
+            logger.trace('{} listener event: {}'.format(self, event))
+            if event['event'] == zmq.EVENT_ACCEPTED:
+                self.connected_clients_num += 1
+            if event['event'] == zmq.EVENT_DISCONNECTED:
+                assert self.connected_clients_num > 0
```

Comment:
asserts will be dropped in production. Should we use an exception here? Or it can be expected that we get disconnected event without connected one?

Reply:
I've added a condition not to get negative numbers in addition to assertion. Also added logging. My concern here is about lost ZMQ events and I don't think that exception is needed here.